aiida_restapi/routers/nodes.py

```diff
@@ -0,0 +1,97 @@
+# -*- coding: utf-8 -*-
+"""Declaration of FastAPI application."""
+from typing import List, Optional
+
+from aiida import orm
+from aiida.cmdline.utils.decorators import with_dbenv
+from fastapi import APIRouter, Depends, File, HTTPException
+from importlib_metadata import entry_points
+
+from aiida_restapi import models
+
+from .auth import get_current_active_user
+
+router = APIRouter()
+
+ENTRY_POINTS = entry_points()
+
+
+@router.get("/nodes", response_model=List[models.Node])
+@with_dbenv()
+async def read_nodes() -> List[models.Node]:
+    """Get list of all nodes"""
+    return models.Node.get_entities()
+
+
+@router.get("/nodes/projectable_properties", response_model=List[str])
+async def get_nodes_projectable_properties() -> List[str]:
+    """Get projectable properties for nodes endpoint"""
+
+    return models.Node.get_projectable_properties()
+
+
+@router.get("/nodes/{nodes_id}", response_model=models.Node)
+@with_dbenv()
+async def read_node(nodes_id: int) -> Optional[models.Node]:
+    """Get nodes by id."""
+    qbobj = orm.QueryBuilder()
+
+    qbobj.append(orm.Node, filters={"id": nodes_id}, project=["**"], tag="node").limit(
+        1
+    )
+    return qbobj.dict()[0]["node"]
+
+
+@router.post("/nodes", response_model=models.Node)
+@with_dbenv()
+async def create_node(
+    node: models.Node_Post,
+    current_user: models.User = Depends(
+        get_current_active_user
+    ),  # pylint: disable=unused-argument
+) -> models.Node:
+    """Create new AiiDA node."""
+
+    node_dict = node.dict(exclude_unset=True)
+    node_type = node_dict.pop("node_type", None)
+
+    try:
+        (entry_point_node,) = ENTRY_POINTS.select(
+            group="aiida.rest.post", name=node_type
+        )
+    except ValueError as exc:
+        raise HTTPException(
+            status_code=404, detail="Entry point '{}' not recognized.".format(node_type)
+        ) from exc
+
+    try:
+        orm_object = entry_point_node.load().create_new_node(node_type, node_dict)
+    except (TypeError, ValueError, KeyError) as err:
+        raise HTTPException(status_code=400, detail="Error: {0}".format(err)) from err
+
+    return models.Node.from_orm(orm_object)
+
+
+@router.post("/singlefiledata", response_model=models.Node)
+@with_dbenv()
+async def create_upload_file(
+    upload_file: bytes = File(...),
+    params: models.Node_Post = Depends(models.Node_Post.as_form),  # type: ignore # pylint: disable=maybe-no-member
+    current_user: models.User = Depends(
+        get_current_active_user
+    ),  # pylint: disable=unused-argument
+) -> models.Node:
+    """Test function for upload file will be merged with create_node later."""
```

Comment:
Ah hm... while writing the commit message, I realized that there is still this special endpoint for singlefiledata.

As you mention here, this should be merged with the `/nodes` endpoint (we don't want to have a separate endpoint for every node type that requires a file).

@NinadBhat  Can you please take care of this?
After this, I think this PR is ready to merge.

Reply:
@ltalirz  I thought we had decided to have two separate endpoints one for normal node and one for file upload. As one uses JSON data and the other uses form data.

Reply:
thanks for the reminder @NinadBhat . After syncing with @chrisjsewell we agree to keep your approach but to move the endpoint under `/nodes/singlefile`.

Reply:
@ltalirz Done